sys/net/credman/credman.c

```diff
@@ -0,0 +1,167 @@
+            DEBUG("credman: invalid PSK parameters\n");
+            ret = CREDMAN_INVALID;
+            goto end;
+        }
+        break;
+    case CREDMAN_TYPE_ECDSA:
+        if ((credential->params.ecdsa.private_key == NULL) ||
+            (credential->params.ecdsa.public_key.x == NULL) ||
+            (credential->params.ecdsa.public_key.y == NULL)) {
+            DEBUG("credman: invalid ECDSA parameters\n");
+            ret = CREDMAN_INVALID;
+            goto end;
+        }
+        break;
+    default:
+        ret = CREDMAN_TYPE_UNKNOWN;
+        goto end;
+    }
+    pos = _find_credential_pos(credential->tag, credential->type);
+    if (pos >= 0) {
+        DEBUG("credman: credential with tag %d and type %d already exist\n",
+              credential->tag, credential->type);
+        ret = CREDMAN_EXIST;
+        goto end;
+    }
+    pos = _find_next_free_pos();
+    if (pos < 0) {
+        DEBUG("credman: no space for new credential\n");
+        ret = CREDMAN_NO_SPACE;
+        goto end;
+    }
+
+    memcpy(&credentials[pos], credential, sizeof(*credential));
+    used++;
+    ret = CREDMAN_OK;
+end:
+    mutex_unlock(&_mutex);
+    return ret;
+}
+
+int credman_get(credman_credential_t *credential,credman_tag_t tag,
+                credman_type_t type)
+{
+    assert(credential);
+    mutex_lock(&_mutex);
+    int ret = CREDMAN_ERROR;
+
+    int pos = _find_credential_pos(tag, type);
+    if (pos < 0) {
+        DEBUG("credman: credential with tag %d and type %d not found\n",
+               tag, type);
+        ret = CREDMAN_NOT_FOUND;
+        goto end;
+    }
+    memcpy(credential, &credentials[pos], sizeof(credman_credential_t));
+    ret = CREDMAN_OK;
+end:
+    mutex_unlock(&_mutex);
+    return ret;
+}
+
+int credman_delete(credman_tag_t tag, credman_type_t type)
+{
+    mutex_lock(&_mutex);
+    int ret = CREDMAN_ERROR;
+    int pos = _find_credential_pos(tag, type);
+    if (pos < 0) {
+        DEBUG("credman: credential with tag %d and type %d not found\n",
+               tag, type);
+        ret = CREDMAN_NOT_FOUND;
+        goto end;
+    }
+    memset(&credentials[pos], 0, sizeof(credman_credential_t));
+    used--;
+    ret = CREDMAN_OK;
+end:
+    mutex_unlock(&_mutex);
+    return ret;
+}
+
+int credman_get_used_count(void)
+{
+    return used;
+}
+
+static int _find_credential_pos(credman_tag_t tag, credman_type_t type)
+{
+    for (unsigned i = 0; i < CREDMAN_MAX_CREDENTIALS; i++) {
+        credman_credential_t *c = &credentials[i];
+        if ((c->tag == tag) && (c->type == type)) {
+            return i;
+        }
+    }
+    return -1;
+}
+
+static int _find_next_free_pos(void)
+{
+    for (int i = 0; i < CREDMAN_MAX_CREDENTIALS; i++) {
+        credman_credential_t *c = &credentials[i];
+        if ((c->type == CREDMAN_TYPE_EMPTY) &&
+            (c->tag == CREDMAN_TAG_EMPTY)) {
+            return i;
+        }
+    }
+    return (used == 0) ? 0 : -1;
+}
```

Comment:
Can be simplified to

```C
static int _find_next_free_pos(void) {
    int i = _find_credential_pos(CREDMAN_TYPE_EMPTY, CREDMAN_TAG_EMPTY);
    if (i < 0) {
        return (used == 0) ? 0 : -1;
    }
    else {
        return i;
    }
}
```

It also reduces code duplication.

Reply:
Also, the `(used == 0) ? 0 : -1` check is redundant, since `(used == 0)` can only be true if all entries are empty and thus also `credentials[0]`.